tinygrad/codegen/kernel.py

```diff
@@ -389,8 +389,21 @@ def apply_opt(self, opt:Opt, append_opt:bool=True):
     if self.reduceop and (opt.op in {OptOps.GROUP, OptOps.GROUPTOP} or (self.group_for_reduces and opt.op not in {OptOps.NOLOCALS, OptOps.PADTO})):
       acc_sz = self.reduceop.dtype.itemsize
       upcast_sz = prod([a for a,b in zip(self.full_shape[self.first_upcast:], self.sts[0].shape[self.first_upcast:]) if a == b])
-      local_sz = prod(self.full_shape[self.first_reduce-self.local_dims:self.first_reduce+self.group_for_reduces])
-      smem_sz = amt*acc_sz*upcast_sz*local_sz
+      local_sz = prod(self.full_shape[self.first_reduce-self.local_dims:self.first_reduce])
+
+      # simplest way to do the sum imo
+      grouped_sz = 0
+      for sh0,sh1 in [(self.sts[r*2+len(self.bufs)].shape, self.sts[r*2+len(self.bufs)+1].shape) for r in range(len(self.reduceops))]:
+        reduce_sz = prod(s1 for s0,s1 in list(zip(sh0, sh1))[self.first_reduce:self.first_reduce+self.group_for_reduces] if s0 != s1)
+        if sh0[axis] != sh1[axis]: reduce_sz *= amt
+        grouped_sz += reduce_sz
+
+      # another way to do it in fewer lines but a bit less clear
+      # grouped_sz = sum(prod(s1 for s0,s1 in list(zip(sh0, sh1))[self.first_reduce:self.first_reduce+self.group_for_reduces] if s0 != s1) \
+      #   * (amt if sh0[axis] != sh1[axis] else 1) \
+      #   for sh0,sh1 in [(self.sts[r*2+len(self.bufs)].shape, self.sts[r*2+len(self.bufs)+1].shape) for r in range(len(self.reduceops))])
```

Comment:
this is another way to calculate the smem usage that saves lines but is a little less clear